# Read the rasters and get their transforms

In [48]:
import os
import glob
from osgeo import gdal
from osgeo import ogr

In [57]:
def GetRasterExtent(gtif):
    geoTransform = gtif.GetGeoTransform()
    minx = geoTransform[0]
    maxy = geoTransform[3]
    maxx = minx + geoTransform[1] * gtif.RasterXSize
    miny = maxy + geoTransform[5] * gtif.RasterYSize
    return minx, miny, maxx, maxy
#Ref: https://gis.stackexchange.com/questions/104362/how-to-get-extent-out-of-geotiff

In [58]:
def GetRasterStats(path):
    listing = glob.glob(path + '*TIF')
    tifFiles = []
    for filename in listing:
        tifFiles.append(filename)
    extents = []
    rasterCounts = []
    sizes = []
    for file in tifFiles:
        gtif = gdal.Open(file)
        extents.append(GetRasterExtent(gtif))
        rasterCounts.append(gtif.RasterCount)
        sizes.append((gtif.RasterXSize, gtif.RasterYSize))
    return tifFiles, extents, rasterCounts, sizes

In [69]:
EO1RasterPath='/home/sravya/data/satellite/haiti/raster/EO1/'
tifFiles, EO1extents, rasterCounts, sizes = GetRasterStats(EO1RasterPath)

In [60]:
tifFiles

['/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B04_L1T.TIF',
 '/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B10_L1T.TIF',
 '/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B02_L1T.TIF',
 '/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B07_L1T.TIF',
 '/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B09_L1T.TIF',
 '/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B05_L1T.TIF',
 '/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B03_L1T.TIF',
 '/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B08_L1T.TIF',
 '/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B01_L1T.TIF',
 '/home/sravya/data/satellite/haiti/raster/EO1/EO1A0090472010015110P0_B06_L1T.TIF']

In [70]:
EO1extents

[(751785.0, 1972785.0, 812715.0, 2111415.0),
 (751785.0, 1972785.0, 812715.0, 2111415.0),
 (751785.0, 1972785.0, 812715.0, 2111415.0),
 (751785.0, 1972785.0, 812715.0, 2111415.0),
 (751785.0, 1972785.0, 812715.0, 2111415.0),
 (751785.0, 1972785.0, 812715.0, 2111415.0),
 (751785.0, 1972785.0, 812715.0, 2111415.0),
 (751785.0, 1972785.0, 812715.0, 2111415.0),
 (751795.0, 1972795.0, 812705.0, 2111405.0),
 (751785.0, 1972785.0, 812715.0, 2111415.0)]

In [62]:
rasterCounts

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]

In [63]:
sizes

[(2031, 4621),
 (2031, 4621),
 (2031, 4621),
 (2031, 4621),
 (2031, 4621),
 (2031, 4621),
 (2031, 4621),
 (2031, 4621),
 (6091, 13861),
 (2031, 4621)]

In [71]:
LandSatRasterPath='/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/'
tifFiles, LANDextents, rasterCounts, sizes = GetRasterStats(LandSatRasterPath)

In [65]:
tifFiles

['/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/LS070080472010011400000000MS00_B01.TIF',
 '/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/LS070080472010011400000000MS00_B62.TIF',
 '/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/LS070080472010011400000000MS00_B61.TIF',
 '/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/LS070080472010011400000000MS00_B02.TIF',
 '/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/LS070080472010011400000000MS00_B08.TIF',
 '/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/LS070080472010011400000000MS00_B05.TIF',
 '/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/LS070080472010011400000000MS00_B04.TIF',
 '/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/LS070080472010011400000000MS00_B07.TIF',
 '/home/sravya/data/satellite/haiti/raster/LANDSAT_ETM+_USGS/LS070080472010011400000000MS00_B03.TIF']

In [72]:
LANDextents

[(131984.99999999997, 1972185.0, 382215.0, 2185815.0),
 (131969.99999999997, 1972170.0, 382230.0, 2185830.0),
 (131969.99999999997, 1972170.0, 382230.0, 2185830.0),
 (131984.99999999997, 1972185.0, 382215.0, 2185815.0),
 (131992.49999999997, 1972192.5, 382207.5, 2185807.5),
 (131984.99999999997, 1972185.0, 382215.0, 2185815.0),
 (131984.99999999997, 1972185.0, 382215.0, 2185815.0),
 (131984.99999999997, 1972185.0, 382215.0, 2185815.0),
 (131984.99999999997, 1972185.0, 382215.0, 2185815.0)]

In [44]:
rasterCounts

[1, 1, 1, 1, 1, 1, 1, 1, 1]

In [45]:
sizes

[(8341, 7121),
 (4171, 3561),
 (4171, 3561),
 (8341, 7121),
 (16681, 14241),
 (8341, 7121),
 (8341, 7121),
 (8341, 7121),
 (8341, 7121)]

# Key insights
1. Seems like all tiffs in a bundle are overlapping -> same geatransform and extent
2. They are all monochromatic -> single band
3. Landsat ones are bigger in terms of pixels.
4. EO1 extent=(751785.0, 1972785.0, 812715.0, 2111415.0)
5. LandSat extent=(131984.99999999997, 1972185.0, 382215.0, 2185815.0) 

# Shape files

In [50]:
import os
shapefile = '/home/sravya/data/satellite/haiti/shapefile/'
driver = ogr.GetDriverByName('ESRI Shapefile')
shp_ds = driver.Open(shapefile, 0) # 0 means read-only. 1 means writeable.

# Check to see if shapefile is found.
if shp_ds is None:
    print("Could not open %s" % (shp_ds))
else:
    print('Opened %s' % (shp_ds))
    layer = shp_ds.GetLayer()
    featureCount = layer.GetFeatureCount()
    print("Number of features : %d" % (featureCount))

Opened <osgeo.ogr.DataSource; proxy of <Swig Object of type 'OGRDataSourceShadow *' at 0x7fea84eb9420> >
Number of features : 294170


In [51]:
layer.GetDescription(), layer.GetExtent()

('PDNA_HTI_2010_Atlas_of_Building_Damage_Assessment_UNOSAT_JRC_WB_v2',
 (724423.8898999998, 814769.2341999998, 2015975.4824, 2119997.0555000007))

In [53]:
feat0=layer.GetFeature(0)
feat0.ExportToJson()

'{"type": "Feature", "geometry": {"type": "Point", "coordinates": [780457.6909999996, 2053277.8935]}, "properties": {"OBJECTID": 1, "SensorDate": "2010/01/21", "ID_Com": 111, "NOM_SECTIO": "6\\u00e8me Turgeau", "Id_dep": 1, "id_Section": "111-01", "Departemen": "OUEST", "Commune": "PORT-AU-PRINCE", "No_section": 6, "Section_": "TURGEAU", "Population": 14760, "Site_ID": "Building (General)", "Damage_ID": "GRADE 4 Very Heavy Damage", "Sensor_ID": "Aerial Photography (Google)", "Analysis": "UNITAR/UNOSAT", "Validation": "Not yet field validated", "Landcover": "High Density Built-Up Zone"}, "id": 0}'

In [54]:
damage_ids=[]
for feature in layer:
    damage_ids.append(feature.GetField("Damage_ID"))

In [55]:
from collections import Counter
Counter(damage_ids)

Counter({'GRADE 1 No Visible Damage': 222253,
         'GRADE 3 Substanial to Heavy Damage': 15737,
         'GRADE 4 Very Heavy Damage': 24677,
         'GRADE 5 Destruction': 31489,
         'Possible Damage': 14})

In [56]:
geom=feat0.GetGeometryRef()
geom.GetX(), geom.GetY()

(780457.6909999996, 2053277.8935000002)

In [67]:
geom_names=[]
for i in range(0,layer.GetFeatureCount()):
    feat = layer.GetFeature(i)
    geom = feat.GetGeometryRef()
    geom_names.append(geom.GetGeometryName())
Counter(geom_names)

Counter({'POINT': 294170})

In [73]:
#xy[0]=xmin
#xy[1]=ymin
                #xy[2]=xmax
                #xy[3]=ymax
def setSpatialFilter(layer, xy):
    wkt = "POLYGON (({} {},{} {},{} {},{} {}))".format(xy[0],xy[1],xy[2],xy[1],xy[2],xy[3],xy[0],xy[3])
    print(wkt)
    layer.SetSpatialFilter(ogr.CreateGeometryFromWkt(wkt))

In [82]:
setSpatialFilter(layer, EO1extents[0])

POLYGON ((751785.0 1972785.0,812715.0 1972785.0,812715.0 2111415.0,751785.0 2111415.0))


In [83]:
layer.GetFeatureCount()

278619

In [84]:
layer.ResetReading()
setSpatialFilter(layer, LANDextents[0])

POLYGON ((131984.99999999997 1972185.0,382215.0 1972185.0,382215.0 2185815.0,131984.99999999997 2185815.0))


In [85]:
layer.GetFeatureCount()

0

# Key insights
1. Labels are points not polygons
2. We have 294170 labels in total
3. Overlap between EO1 raster is 278619
4. Overlap between LandSat raster is 0

# TODO: Visualize points over EO1 tiff